core/src/main/java/hudson/model/Queue.java

```diff
@@ -2317,6 +2353,71 @@ public CauseOfBlockage getCauseOfBlockage() {
     }
 
     /**
+     * {@link Item} in the {@link Queue#flyWeightTasks} stage.
+     */
+    public final class FlyWeightItem extends NotWaitingItem {
```

Comment:
Or you could create a base class for the shared functionality in order to simplify the things